packages/plugins/apollo-usage-report/src/index.ts

```diff
@@ -0,0 +1,172 @@
+import { defaultUsageReportingSignature } from 'apollo-graphql';
+import { Report } from 'apollo-reporting-protobuf';
+import { printSchema } from 'graphql';
+import { isAsyncIterable, Plugin, YogaLogger } from 'graphql-yoga';
+import {
+  ApolloInlineTraceContext,
+  ApolloInlineTracePluginOptions,
+  useApolloInstrumentation,
+} from '@graphql-yoga/plugin-apollo-inline-trace';
+import { fetch } from '@whatwg-node/fetch';
+
+type ApolloUsageReportOptions = ApolloInlineTracePluginOptions & {
+  /**
+   * The graph ref of the managed federation graph.
+   * It is composed of the graph ID and the variant (`<YOUR_GRAPH_ID>@<VARIANT>`).
+   *
+   * If not provided, `APOLLO_GRAPH_REF` environment variable is used.
+   *
+   * You can find a a graph's ref at the top of its Schema Reference page in Apollo Studio.
+   */
+  graphRef?: string;
+  /**
+   * The API key to use to authenticate with the managed federation up link.
+   * It needs at least the `service:read` permission.
+   *
+   * If not provided, `APOLLO_KEY` environment variable will be used instead.
+   *
+   * [Learn how to create an API key](https://www.apollographql.com/docs/federation/v1/managed-federation/setup#4-connect-the-gateway-to-studio)
+   */
+  apiKey?: string;
+  /**
+   * Usage report endpoint
+   *
+   * Defaults to GraphOS endpoint (https://usage-reporting.api.apollographql.com/api/ingress/traces)
+   */
+  endpoint?: string;
+};
+
+const DEFAULT_REPORTING_ENDPOINT =
+  'https://usage-reporting.api.apollographql.com/api/ingress/traces';
+
+export function useApolloUsageReport(options: ApolloUsageReportOptions = {}): Plugin {
+  const [instrumentation, ctxForReq] = useApolloInstrumentation(options) as [
+    Plugin,
+    WeakMap<
+      Request,
+      ApolloInlineTraceContext & {
+        schemaId?: string;
+        operationKey?: string;
+      }
+    >,
+  ];
+
+  const {
+    graphRef = process.env['APOLLO_GRAPH_REF'],
+    apiKey = process.env['APOLLO_KEY'],
+    endpoint = DEFAULT_REPORTING_ENDPOINT,
+  } = options;
+
+  let logger: YogaLogger;
+  let schemaId: string;
+
+  return {
+    onPluginInit({ addPlugin }) {
+      addPlugin(instrumentation);
+      addPlugin({
+        async onYogaInit(args) {
+          logger = Object.fromEntries(
```

Comment:
This might lead some problems with class instances?
Maybe we can use the prefix inside the functions instead.

Reply:
Yes probably, I've fixed it. We should definitly have the same thing than Mesh for loggers: and `createChild` method

Reply:
Yes, if we implement it, we don't need an additional impl in Mesh too